workflow last job info

prime-time opened the pull request:
The hubble/repairs page is slow to load when there are a large number of active workflows. The reason for the slowness is that it iterates through every active workflow and loads the full workflow history for each one. We are in the process of modifying Hubble to query Elasticsearch for active and recent workflows instead of using workflow-manager, and we are also exploring the possibility of querying Elasticsearch for failed workflows. This change introduces a new method that fetches the most recent events in the state machine's execution history, and stores the last Job info in the `WorkflowSummary`. The new method shares a lot of logic with https://github.com/Clever/workflow-manager/blob/f0f8abee201526cbe7f51ee93fcb41dbf6d0cd27/executor/workflow_manager_sfn.go#L372, and the main difference that it doesn't page through the entire execution history to fill out the Jobs array and instead just looks at the most recent events. Some of the code from `UpdateWorkflowHistory` is refactored into separate functions so that they can be used in the new `updateWorkflowLastJob` method. 

Here's an earlier draft at this that would fill in the entire jobs array https://github.com/Clever/workflow-manager/pull/170. The downside of that approach is that it would call `GetExecutionHistory` for each failed workflow update, and that endpoint has a relatively low rate limit.

Below is an example workflow that I created with `lastJob` populated
```
[renatoprime@Renatos-MacBook-Pro workflow-manager (SYNC-1172-workflow-last-attempted-state-v2)]$ curl https://vpc-workflows-dev-2nfzpvqce62oumm6zcnjaipife.us-west-1.es.amazonaws.com/clever-dev-workflow-manager-dev-v3-workflows/default/fa275130-4be8-42be-85ac-7a6cdf378268 | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2055  100  2055    0     0  17471      0 --:--:-- --:--:-- --:--:-- 17415
{
  "_index": "clever-dev-workflow-manager-dev-v3-workflows",
  "_type": "default",
  "_id": "fa275130-4be8-42be-85ac-7a6cdf378268",
  "_version": 3,
  "found": true,
  "_source": {
    "Workflow": {
      "createdAt": "2019-04-23T15:00:53.795429Z",
      "id": "fa275130-4be8-42be-85ac-7a6cdf378268",
      "input": "{\"uh\":\"boo\"}",
      "lastJob": {
        "createdAt": "2019-04-23T15:00:54Z",
        "id": "2",
        "input": "{\"_EXECUTION_NAME\":\"fa275130-4be8-42be-85ac-7a6cdf378268\",\"uh\":\"boo\"}",
        "startedAt": "2019-04-23T15:00:54Z",
        "state": "clogger",
        "stateResource": {
          "lastUpdated": "2019-04-23T15:00:54Z",
          "name": "clogger",
          "namespace": "clever-dev",
          "type": "LambdaFunctionARN"
        },
        "status": "failed",
        "statusReason": "{\"errorMessage\":\"🐖  🐖  🐖  🐖  🐖  🐖\",\"errorType\":\"errorString\"}\nerrorString",
        "stoppedAt": "2019-04-23T15:00:54Z"
      },
      "lastUpdated": "2019-04-23T15:01:54.667503Z",
      "namespace": "clever-dev",
      "queue": "default",
      "status": "failed",
      "stoppedAt": "2019-04-23T15:00:54Z",
      "tags": {
        "team": "eng-secure-sync"
      },
      "workflowDefinition": {
        "createdAt": "2019-04-11T23:50:49.122636234Z",
        "defaultTags": {
          "team": "eng-secure-sync"
        },
        "id": "442052ba-64c0-4119-a1eb-4f4ff4dad16a",
        "manager": "step-functions",
        "name": "clogger:back-it-up",
        "stateMachine": {
          "StartAt": "clogger",
          "States": {
            "clogger": {
              "HeartbeatSeconds": "30",
              "Next": "dpt-clogger",
              "Resource": "lambda:clogger",
              "Retry": [
                {
                  "ErrorEquals": [
                    "States.ALL"
                  ],
                  "MaxAttempts": "0"
                }
              ],
              "TimeoutSeconds": "7200",
              "Type": "Task"
            },
            "dpt-clogger": {
              "HeartbeatSeconds": "30",
              "Next": "mpt-clogger",
              "Resource": "lambda:dpt-clogger",
              "Retry": [
                {
                  "ErrorEquals": [
                    "States.ALL"
                  ],
                  "MaxAttempts": "0"
                }
              ],
              "TimeoutSeconds": "7200",
              "Type": "Task"
            },
            "mpt-clogger": {
              "End": true,
              "HeartbeatSeconds": "30",
              "Resource": "lambda:mpt-clogger",
              "Retry": [
                {
                  "ErrorEquals": [
                    "States.ALL"
                  ],
                  "MaxAttempts": "0"
                }
              ],
              "TimeoutSeconds": "7200",
              "Type": "Task"
            }
          },
          "TimeoutSeconds": "259200",
          "Version": "1.0"
        },
        "version": "3"
      }
    },
    "__ttl": "1558623653",
    "_gsi-ca": "2019-04-23T15:00:53.795429Z",
    "_gsi-wn": "clogger:back-it-up",
    "_gsi-wn-and-resolvedbyuser": "clogger:back-it-up:false",
    "_gsi-wn-and-status": "clogger:back-it-up:failed",
    "id": "fa275130-4be8-42be-85ac-7a6cdf378268"
  }
}
```

- [x] Update swagger.yml version
- [x] Run "make generate"

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -592,6 +592,246 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
+
+	jobs := []*models.Job{}
+	eventIDToJob := map[int64]*models.Job{}
+	eventToJob := func(evt *sfn.HistoryEvent) *models.Job {
```
can we extract this into its own function rather than copy it?

prime-time replied:
yeah I think so

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -592,6 +592,246 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
 	return wm.store.UpdateWorkflow(ctx, *workflow)
 }
 
+// UpdateWorkflowLastJob queries AWS step-functions for the latest in the event history for the given state machine,
```
I think it'd make sense to add a section to the comment about why we track just the last job and not the full execution history for failed jobs. The context for this change could get lost.

prime-time replied:
Good point. I'll add in a note.

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -592,6 +592,246 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
+		}
+	}
+
+	lastJobEventID := aws.Int64Value(historyOutput.Events[lastJobIndex].Id)
```
it's basically just these two lines that differentiate from the copied function? I have to believe this can be made more DRY?

prime-time replied:
This pull request is kinda drafty. (It's not marked as such because I thought opening a "draft" pull request would reduce the chances of getting any early eyes on it). At the time of opening this, we weren't 100% confident that we would implement this approach.

prime-time commented:
I'm wondering if we should put the new code that populates `LastJob` inside of `UpdateWorkflowSummary` if we're keeping `LastJob` in the `WorkflowSummary` object. We could have it still only populate `LastJob` on failures.
https://github.com/Clever/workflow-manager/blob/5a9490c4881822915979e272c797e43e03929793/executor/workflow_manager_sfn.go#L357-L369

prime-time commented:
Thanks for taking a look at the draft. I incorporated suggestions about extracting the shared code and added more to the comment that explains the added method. I rewrote the commits, and I tried to organize them so that it would be easier to review the diffs by commit. 

I also tested it again to verify that `lastJob` gets populated with the changes
```
{
  "_index": "clever-dev-workflow-manager-dev-v3-workflows",
  "_type": "default",
  "_id": "169a2c4b-27d4-4df3-8fa4-3eb7eb1f4a11",
  "_version": 2,
  "found": true,
  "_source": {
    "Workflow": {
      "createdAt": "2019-04-27T22:38:50.883024Z",
      "id": "169a2c4b-27d4-4df3-8fa4-3eb7eb1f4a11",
      "input": "{}",
      "lastJob": {
        "createdAt": "2019-04-27T22:38:51Z",
        "id": "2",
        "input": "{\"_EXECUTION_NAME\":\"169a2c4b-27d4-4df3-8fa4-3eb7eb1f4a11\"}",
        "startedAt": "0001-01-01T00:00:00Z",
        "state": "clogger",
        "stateResource": {
          "lastUpdated": "2019-04-27T22:38:51Z",
          "name": "clogger",
          "type": "ActivityARN"
        },
        "status": "failed",
        "statusReason": "An error occurred while executing the state 'clogger' (entered at the event id #2). The activity arn:aws:states:us-west-2:589690932525:activity:--clogger does not exist.\nStates.Runtime",
        "stoppedAt": "2019-04-27T22:38:51Z"
      },
      "lastUpdated": "2019-04-27T22:39:51.666323Z",
      "queue": "default",
      "status": "failed",
      "stoppedAt": "2019-04-27T22:38:51Z",
      "tags": {
        "team": "eng-secure-sync"
      },
      "workflowDefinition": {
        "createdAt": "2019-04-10T22:47:12.925692265Z",
        "defaultTags": {
          "team": "eng-secure-sync"
        },
        "id": "fc1409d8-cf7e-432f-9a39-4c4d75cdf3e2",
        "manager": "step-functions",
        "name": "clogger:back-it-up",
        "stateMachine": {
          "StartAt": "clogger",
          "States": {
            "clogger": {
              "HeartbeatSeconds": "30",
              "Next": "dpt-clogger",
              "Resource": "clogger",
              "Retry": [
                {
                  "ErrorEquals": [
                    "States.ALL"
                  ],
                  "MaxAttempts": "0"
                }
              ],
              "TimeoutSeconds": "7200",
              "Type": "Task"
            },
            "dpt-clogger": {
              "HeartbeatSeconds": "30",
              "Next": "mpt-clogger",
              "Resource": "dpt-clogger",
              "Retry": [
                {
                  "ErrorEquals": [
                    "States.ALL"
                  ],
                  "MaxAttempts": "0"
                }
              ],
              "TimeoutSeconds": "7200",
              "Type": "Task"
            },
            "mpt-clogger": {
              "End": true,
              "HeartbeatSeconds": "30",
              "Resource": "mpt-clogger",
              "Retry": [
                {
                  "ErrorEquals": [
                    "States.ALL"
                  ],
                  "MaxAttempts": "0"
                }
              ],
              "TimeoutSeconds": "7200",
              "Type": "Task"
            }
          },
          "TimeoutSeconds": "259200",
          "Version": "1.0"
        }
      }
    },
    "__ttl": "1558996730",
    "_gsi-ca": "2019-04-27T22:38:50.883024Z",
    "_gsi-wn": "clogger:back-it-up",
    "id": "169a2c4b-27d4-4df3-8fa4-3eb7eb1f4a11"
  }
}
```

prime-time commented on executor/workflow_manager_sfn.go:
```diff
@@ -410,6 +410,76 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
+
+	jobs := []*models.Job{}
+	eventIDToJob := map[int64]*models.Job{}
+	eventToJobFn := newEventToJobFn(execARN, eventIDToJob, &jobs)
```
Here is the `eventToJobFn` logic that was extracted from `UpdateWorkflowHistory`

prime-time commented on executor/workflow_manager_sfn.go:
```diff
@@ -410,6 +410,76 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
+		orderedLatestEvents = append(orderedLatestEvents, historyOutput.Events[lastJobIndex-i])
+	}
+
+	processEvents(orderedLatestEvents, eventToJobFn, workflow)
```
Here is more logic that was extracted from `UpdateWorkflowHistory`

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -650,3 +507,154 @@ func newEventToJobFn(
+	eventToJobFn func(evt *sfn.HistoryEvent) *models.Job,
+	workflow *models.Workflow) {
+	for _, evt := range events {
+		job := eventToJobFn(evt)
```
I know you're just moving this code around..... but wow this was hard to figure out.
So `newEventToJobFn` is appending to the `jobs` array that was passed to it when it was constructed?

There's gotta be a more intuitive way to express this... 🤔

prime-time replied:
🙈

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -410,6 +410,76 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
+// and it uses these events to populate LastJob within the WorkflowSummary.
+// This method is heavily inspired by UpdateWorkflowHistory, however updateWorkflowLastJob only
+// finds the last Job which is all that is needed to determine the state where a workflow has failed.
+// We try to use the sfn GetExecutionHistory API endpoint somewhat sparingly because
```
appreciate all the context here 👍

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -410,6 +410,76 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
+		}
+	}
+	if lastJobIndex < 0 {
+		log.InfoD("last-job-not-found", logger.M{
```
should this be an error log?

prime-time replied:
I'll change it to ErrorD

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -364,6 +364,11 @@ func (wm *SFNWorkflowManager) UpdateWorkflowSummary(ctx context.Context, workflo
 	if workflow.Status == models.WorkflowStatusSucceeded || workflow.Status == models.WorkflowStatusCancelled {
 		workflow.ResolvedByUser = true
 	}
+	if workflow.Status == models.WorkflowStatusFailed {
```
comment on this block?

prime-time replied:
okie dokie

natebrennand commented on executor/workflow_manager_sfn_test.go:
```diff
@@ -1087,9 +1087,6 @@ func TestUpdateWorkflowStatusWorkflowTimedOut(t *testing.T) {
 	require.NoError(t, c.manager.UpdateWorkflowSummary(ctx, workflow))
 	require.NoError(t, c.manager.UpdateWorkflowHistory(ctx, workflow))
 	assert.Equal(t, models.WorkflowStatusFailed, workflow.Status)
-	// TODO: this line below is not testing anything. Verify what needs to go in this assertion
```
:wat:

prime-time commented on executor/workflow_manager_sfn.go:
```diff
@@ -641,3 +602,51 @@ func causeAndErrorNameFromFailureEvent(evt *sfn.HistoryEvent) (string, string) {
 }
+
+func newEventToJobFn(
+	execARN string, eventIDToJob map[int64]*models.Job, jobsPtr *[]*models.Job,
```
I'm actually using it in one place outside the function to map the event id of the last job to the job, but I can try changing things around so that it can be scoped within the function.

prime-time commented:
Sorry for the earlier headache. I cowardly chose to modify the existing code the least amount possible, but Nate inspired me to try and refactor some things to make it easier to follow.

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -390,24 +390,22 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
 		// 2) set `reverseOrder` to true to get most recent events first
 		// 3) stop paging once we get to to the smallest job ID (aka event ID) that is still pending
-		processEvents(historyOutput.Events, eventToJobFn, workflow)
+		jobs = processEvents(historyOutput.Events, execARN, workflow)
```
should we be appending the results here? Seems like this callback might be executed more than once

prime-time replied:
good catch ⚾️

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -537,60 +530,68 @@ func causeAndErrorNameFromFailureEvent(evt *sfn.HistoryEvent) (string, string) {
+	switch *evt.Type {
+	case sfn.HistoryEventTypeExecutionStarted:
+		// very first event for an execution, so there are no jobs yet
+		job = nil
```
what if we ended every case with a `return`?

prime-time replied:
I removed those when I added `jobs` and `eventIDToJob` as return values, but I'll add them back.

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -537,60 +530,68 @@ func causeAndErrorNameFromFailureEvent(evt *sfn.HistoryEvent) (string, string) {
+	evt *sfn.HistoryEvent,
+	execARN string, jobs []*models.Job,
+	eventIDToJob map[int64]*models.Job,
+) (*models.Job, []*models.Job, map[int64]*models.Job) {
```
`eventToJob` gets mutated by reference throughout this function, idk if we need to necessarily return it
(I generally don't love the model of mutating the maps piecemeal like this 🤷‍♂)

prime-time replied:
I'm still trying to think of a better way to handle this

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -537,60 +530,68 @@ func causeAndErrorNameFromFailureEvent(evt *sfn.HistoryEvent) (string, string) {
-	workflow *models.Workflow) {
+// processEvents constructs the Jobs array from the AWS step-functions event history.
+func processEvents(events []*sfn.HistoryEvent, execARN string, workflow *models.Workflow,
+) []*models.Job {
```
💯

natebrennand commented on executor/workflow_manager_sfn.go:
```diff
@@ -592,6 +415,77 @@ func (wm *SFNWorkflowManager) UpdateWorkflowHistory(ctx context.Context, workflo
+		orderedLatestEvents = append(orderedLatestEvents, historyOutput.Events[lastJobIndex-i])
+	}
+
+	jobs := processEvents(orderedLatestEvents, execARN, workflow)
```
I'm still a little confused why we can't just call `processEvents` and take the last (or first if we still reverse the events) entry. Is it necessary to trim the events down?

prime-time replied:
I did it this way to avoid processing events that occurred before the one of interest, but I realize that the extra work is not significant and what we really want to limit is API calls which we are doing. I'll make the change to simplify this.

rgarcia commented on executor/workflow_manager_sfn_test.go:
```diff
@@ -1065,3 +1132,12 @@ func (c *sfnManagerTestController) updateWorkflow(ctx context.Context, t *testin
+		reversedEvents[i] = events[numEvents-1-i]
+	}
+	return reversedEvents
+}
```
could reverse the slice in-place via
```go
	sort.Slice(events, func(i, j int) bool { return j < i })
```

prime-time replied:
Wow that's so clean. I'll add that to my toolbox. I want to return a new array in the tests though so that I can mock `GetExecutionHistory` when called with and without `reverseOrder`.

prime-time commented:
I did some rebasing on the latest master and combined the latest commits into a single refactor commit. I'm going to do some more manual testing locally. Since this is version < 1, I changed the version bump commit to be a patch instead of a minor version change.
